plugins/devlink/fu-devlink-netlink.c

```diff
@@ -0,0 +1,624 @@
+/*
+ * Copyright 2025 NVIDIA Corporation & Affiliates
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later
+ */
+
+#include "config.h"
+
+#include <glib-unix.h>
+#include <libmnl/libmnl.h>
+#include <linux/genetlink.h>
+#include <linux/netlink.h>
+
+#include "fu-devlink-netlink.h"
+
+static struct nlmsghdr *
+fu_devlink_netlink_msg_prepare(void *buf,
+			       uint32_t nlmsg_type,
+			       gboolean dump,
+			       void *extra_header,
+			       size_t extra_header_size)
+{
+	struct nlmsghdr *nlh;
+	void *eh;
+	g_autoptr(GError) error = NULL;
+
+	nlh = mnl_nlmsg_put_header(buf);
+	nlh->nlmsg_type = nlmsg_type;
+	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK;
+	if (dump)
+		nlh->nlmsg_flags |= NLM_F_DUMP;
+	nlh->nlmsg_seq = time(NULL);
+
+	eh = mnl_nlmsg_put_extra_header(nlh, extra_header_size);
+	memcpy(eh, extra_header, extra_header_size); /* nocheck:blocked */
+
+	return nlh;
+}
+
+/* attribute parser callback for netlink error attributes */
+static int
+fu_devlink_netlink_nlmsgerr_attr_cb(const struct nlattr *attr, void *data)
+{
+	const struct nlattr **tb = data;
+	int type = mnl_attr_get_type(attr);
+
+	if (mnl_attr_type_valid(attr, NLMSGERR_ATTR_MAX) < 0)
+		return MNL_CB_OK;
+
+	tb[type] = attr;
+	return MNL_CB_OK;
+}
+
+typedef struct {
+	void *user_data;
+	mnl_cb_t user_cb;
+	GError **error;
+} FuDevlinkCbHelper;
+
+static gboolean
+fu_devlink_netlink_error_cb_extack(const struct nlmsghdr *nlh, GError **error)
+{
+	const struct nlmsgerr *err = mnl_nlmsg_get_payload(nlh);
+	struct nlattr *tb[NLMSGERR_ATTR_MAX + 1] = {};
+	unsigned int hlen = sizeof(struct nlmsgerr);
+
+	if ((nlh->nlmsg_flags & NLM_F_ACK_TLVS) == 0)
+		return FALSE;
+	if ((nlh->nlmsg_flags & NLM_F_CAPPED) == 0)
+		hlen += mnl_nlmsg_get_payload_len(&err->msg);
+	if (mnl_attr_parse(nlh, hlen, fu_devlink_netlink_nlmsgerr_attr_cb, tb) != MNL_CB_OK)
+		return FALSE;
+	if (tb[NLMSGERR_ATTR_MSG] == NULL)
+		return FALSE;
+	g_set_error(error,
+		    FWUPD_ERROR,
+		    FWUPD_ERROR_NOT_SUPPORTED,
+		    "netlink error: %s (%s)",
```

Comment:
i think we're using `[%s]` for the string version

Reply:
It's a bit misleading. It is an extack message from kernel, it is not string representation of the previous value, it is rather descriptive error message. Perhaps: "netlink error: %s; %s"?